checks.d/haproxy.py

```diff
@@ -319,51 +336,37 @@ def _process_status_metric(self, hosts_statuses, collect_status_metrics_by_host,
-
             if self._is_service_excl_filtered(service, services_incl_filter, services_excl_filter):
                 continue
 
+            tags = []
+            if count_status_by_service:
+                tags.append('service:%s' % service)
             if collect_status_metrics_by_host:
                 tags.append('backend:%s' % hostname)
 
-            self._gauge_all_statuses(
-                "haproxy.count_per_status",
-                count, status, tags, counter
-            )
-
-            if 'up' in status or 'open' in status:
-                agg_statuses[service]['available'] += count
-            if 'down' in status or 'maint' in status or 'nolb' in status:
-                agg_statuses[service]['unavailable'] += count
-
-        if counter is not None:
-            # send aggregated counts as gauges
-            for key, count in counter.iteritems():
-                metric_name, tags = key[0], key[1]
-                self.gauge(metric_name, count, tags=tags)
+            counter_status = status
+            if collate_status_tags_per_host:
+                counter_status = Services.STATUS_MAP.get(status, status)
 
-        for service in agg_statuses:
-            for status, count in agg_statuses[service].iteritems():
-                tags = ['status:%s' % status]
-                if count_status_by_service:
-                    tags.append('service:%s' % service)
-
-                self.gauge("haproxy.count_per_status", count, tags=tags)
+            counter[tuple(tags)][counter_status] += count
 
-    def _gauge_all_statuses(self, metric_name, count, status, tags, counter):
-        if counter is not None:
-            counter_key = tuple([metric_name, tuple(tags + ['status:%s' % status])])
-            counter[counter_key] += count
-        else:
-            # assume we have enough context, just send a gauge
-            self.gauge(metric_name, count, tags + ['status:%s' % status])
-
-        for state in Services.ALL_STATUSES:
-            if state != status:
-                self.gauge(metric_name, 0, tags + ['status:%s' % state.replace(" ", "_")])
+            # Compute aggregates
+            agg_tags = []
+            if count_status_by_service:
+                agg_tags.append('service:%s' % service)
+            agg_statuses_counter[tuple(agg_tags)][Services.STATUS_MAP.get(status, status)] += count
+
+        for key, count_per_status in counter.iteritems():
+            key_tags = list(key)
+            for status, count in count_per_status.iteritems():
+                self.gauge('haproxy.count_per_status', count, tags=key_tags + ['status:%s' % status])
+
+        # Send aggregates
+        for service, service_agg_statuses in agg_statuses_counter.iteritems():
+            service_tags = list(service)
+            for status, count in service_agg_statuses.iteritems():
+                self.gauge("haproxy.count_per_status", count, tags=service_tags + ['status:%s' % status])
 
```

Comment:
Overriding `haproxy.count_per_status` would lead to unexpected behavior.

See
```python
from time import sleep
from statsd import statsd


# Samples
context_a = [u"foo:bar"]
context_b = [u"foo:qux"]

metric_name = u"metric.test.aggregate"


while True:
    # != contexts
    statsd.gauge(metric_name, -10, tags=context_a)
    statsd.gauge(metric_name, 10, tags=context_b)

    # Overwrite aggregate
    statsd.gauge(metric_name, 50, tags=[])

    sleep(10)
```
and
![unspecified](https://cloud.githubusercontent.com/assets/5708089/16019503/4be3dad8-3177-11e6-8113-4e9dd6f3ab42.png)